### PR TITLE
Add first bit of registerDynamically

### PR DESCRIPTION
**Pull request by lukel97**
Needs documented, better names and custom method handling etc. Can do this after this pr

### CODE REVIEW

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs [hunk 1]
```diff
@@ -380,7 +380,19 @@ type family BaseMessage (t :: MethodType) :: Method p t -> Type where
...
 
-type HandlerFunc a = Either ResponseError a -> IO ()
+-- | The type of a handler that handles requests and notifications coming in
+-- from the server or client
+type family Handler m :: Type where
+  Handler (m :: Method p t) = BaseHandler m (IO ())
+
+-- | Version of 'Handler' that can be used to construct arbitrary functions
+-- taking in the required handler arguments
+type family BaseHandler (m :: Method p t) a :: Type where
```
Comment (wz1000) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Handler and BaseHandler seem to be almost the same thing. Do we need both?

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Method.hs [hunk 2]
```diff
@@ -382,3 +383,7 @@ instance A.ToJSON (SMethod m) where
   toJSON SCancelRequest                      = A.String "$/cancelRequest"
 -- Custom
   toJSON (SCustomMethod m)                   = A.String m
+
+instance IxOrd SMethod where
```
Comment (wz1000) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/Method.hs]: This is very unsafe, please remove.

File: src/Language/Haskell/LSP/Core.hs [hunk 3]
```diff
@@ -316,70 +336,77 @@ mkSendReqFunc m params resHandler = do
...
+  -- their handlers
+  sf <- asks resSendMessage
+  dynReqHandlers <- readData resRegistrationsReq -- :: LspM config (RegistrationMap Request)
+  dynNotHandlers <- readData resRegistrationsNot
+  ~() <- case splitClientMethod m of
+    IsClientReq -> 
+      case DMap.lookup m dynReqHandlers of
+        -- TODO: If one request handles it, stop here and break out so we don't send back multiple responses?
+        Just (Compose regs) -> forM_ regs $ \(Pair _regId (RegistrationHandler h)) ->
+          liftIO $ h msg (mkClientResponseCallback sf msg)
```
Comment (wz1000) [inline src/Language/Haskell/LSP/Core.hs]: Do we really want to run `hh` to dispatch the static handlers after the message has been handled by dynamic handlers? I think we should only run `hh` if the dynamic handlers don't run.

Better yet, all of this logic should move to `hh`, as part of the code responsible for the handler lookup.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs (hunk 1 above)
Comment (lukel97) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: I'd like to roll them into one but I'm not sure how to pattern match on both the `m :: Method p t` and the `t :: MethodType` simultaneously in one family, any ideas?

Comment (wz1000) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: but you aren't doing this in `Handler` or `BaseHandler`?

Comment (lukel97) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Huh no idea what I was doing earlier but this seems to work now, we can probably remove `BaseMessage` as well then
```haskell
type family BaseHandler (m :: Method p t) (a :: Type) :: Type where
  BaseHandler (m :: Method p Request)      a = RequestMessage m -> (Either ResponseError (ResponseParams m) -> IO ()) -> a
  BaseHandler (m :: Method p Notification) a = NotificationMessage m -> a
```

**Comment by lukel97** [comment]
@wz1000 should be tidied up now, and I think the best thing to do for now is to throw an error when trying to dynamically register for a custom method as there is no way for the client to advertise that it supports dynamic registration on a custom method.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs [hunk 4]
```diff
@@ -366,6 +366,21 @@ instance FromJSON (ResponseParams a) => FromJSON (ResponseMessage a) where
       (Nothing, Nothing) -> fail "both error and result cannot be Nothing"
     return $ ResponseMessage _jsonrpc _id $ result
 
+-- ---------------------------------------------------------------------
+-- Handlers
+-- ---------------------------------------------------------------------
+
+-- | The type of a handler that handles requests and notifications coming in
+-- from the server or client
+type family Handler m :: Type where
```
Comment (wz1000) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: This does not need to be a type family. Use a type synonym instead

**Pull request merged**